File: plugins/git-utils/src/core/shell.ts
// ============================================================
// Shell — subprocess 실행 유틸리티
// ============================================================

export interface ExecResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

/**
 * 명령어를 실행하고 결과를 반환합니다.
 * 실패 시 throw하지 않고 exitCode를 반환합니다.
 */
export async function exec(
  command: string[],
  options?: { cwd?: string },
): Promise<ExecResult> {
  const proc = Bun.spawn(command, {
    stdout: 'pipe',
    stderr: 'pipe',
    cwd: options?.cwd,
  });

  const [stdout, stderr] = await Promise.all([
    new Response(proc.stdout).text(),
    new Response(proc.stderr).text(),
  ]);

  const exitCode = await proc.exited;

  return {
    stdout: stdout.trimEnd(),
    stderr: stderr.trimEnd(),
    exitCode,
  };
}

/**
 * 명령어를 실행하고, 실패 시 에러를 throw합니다.
 */
export async function execOrThrow(
  command: string[],
  options?: { cwd?: string },
): Promise<string> {
  const result = await exec(command, options);
  if (result.exitCode !== 0) {
    throw new Error(`Command failed (exit ${result.exitCode}): ${command.join(' ')}\n${result.stderr}`);
  }
  return result.stdout;
}

File: plugins/atelier/cli/src/git/core/shell.rs
//! Subprocess execution primitive, ported from git-utils `core/shell.ts`.
//!
//! `exec` never fails the process — it captures stdout/stderr and the exit
//! code so callers can branch on it. `exec_or_throw` is the strict variant
//! that returns an error on a non-zero exit. Both trim trailing whitespace
//! from captured output, matching the TypeScript `trimEnd()` behaviour.

use anyhow::{bail, Result};
use std::path::Path;
use std::process::Command;

/// Captured result of a subprocess run.
#[derive(Debug, Clone)]
pub struct ExecResult {
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
}

/// Runs `command` (argv form, `command[0]` is the program) in `cwd`. Thin
/// wrapper over [`exec_env`] with no extra environment.
pub fn exec(command: &[&str], cwd: Option<&Path>) -> Result<ExecResult> {
    exec_env(command, cwd, &[])
}

/// Like [`exec`] but layers `env` (key/value pairs) on top of the inherited
/// environment. Never returns an error for a non-zero exit — the exit code is
/// surfaced in [`ExecResult::exit_code`].
pub fn exec_env(command: &[&str], cwd: Option<&Path>, env: &[(&str, &str)]) -> Result<ExecResult> {
    let (program, args) = command
        .split_first()
        .ok_or_else(|| anyhow::anyhow!("exec: empty command"))?;
    let mut cmd = Command::new(program);
    cmd.args(args);
    if let Some(dir) = cwd {
        cmd.current_dir(dir);
    }
    for (k, v) in env {
        cmd.env(k, v);
    }
    let output = cmd
        .output()
        .map_err(|e| anyhow::anyhow!("failed to spawn `{program}`: {e}"))?;
    Ok(ExecResult {
        stdout: String::from_utf8_lossy(&output.stdout)
            .trim_end()
            .to_string(),
        stderr: String::from_utf8_lossy(&output.stderr)
            .trim_end()
            .to_string(),
        exit_code: output.status.code().unwrap_or(-1),
    })
}

/// Like [`exec`] but returns an error (with stderr) when the command exits
/// non-zero. On success, returns the trimmed stdout.
pub fn exec_or_throw(command: &[&str], cwd: Option<&Path>) -> Result<String> {
    let result = exec(command, cwd)?;
    if result.exit_code != 0 {
        bail!(
            "Command failed (exit {}): {}\n{}",
            result.exit_code,
            command.join(" "),
            result.stderr
        );
    }
    Ok(result.stdout)
}
